## *`Solution Notebook for Practice Project 2`*

##  *Table of Content*

## **[Part I: Numpy](#np)**
## **[Part II: Pandas](#pd)**

<a id="np"> </a>
# *Part I: NUMPY*

## Q1) State differences between Numpy Array and lists.

 - Arrays are homogeneous where as Lists can be homogeneous or heterogeneous
 - Element wise operation is possible in arrays but not possible in lists
 - Arrays are very useful for multi-dimensional data. 

## Q2) Write a Code to create 50 evenly spaced numbers between 0 to 10.

In [111]:
np.linspace(0,10,50)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

## Q3) Write a Code to create an array of shape (4,5) with random samples with random seed 7.

In [114]:
np.random.seed(7)
arr = np.random.random((4,5))
arr

array([[0.07630829, 0.77991879, 0.43840923, 0.72346518, 0.97798951],
       [0.53849587, 0.50112046, 0.07205113, 0.26843898, 0.4998825 ],
       [0.67923   , 0.80373904, 0.38094113, 0.06593635, 0.2881456 ],
       [0.90959353, 0.21338535, 0.45212396, 0.93120602, 0.02489923]])

## Q4) Write a Code to reshape the array created in Q3 into shape (2,10).

In [119]:
arr.reshape(2,10)

array([[0.07630829, 0.77991879, 0.43840923, 0.72346518, 0.97798951,
        0.53849587, 0.50112046, 0.07205113, 0.26843898, 0.4998825 ],
       [0.67923   , 0.80373904, 0.38094113, 0.06593635, 0.2881456 ,
        0.90959353, 0.21338535, 0.45212396, 0.93120602, 0.02489923]])

## Q5) Create an 1D array starting from 0 ending with 100 with stepsize 10 and another array starting from 1000 ending with 10000 with stepsize 1000 and then concatenate two arrays created.

In [155]:
arr1 = np.arange(10,110,10)
arr1

array([ 10,  20,  30,  40,  50,  60,  70,  80,  90, 100])

In [156]:
arr2 = np.arange(1000,11000,1000)
arr2

array([ 1000,  2000,  3000,  4000,  5000,  6000,  7000,  8000,  9000,
       10000])

In [157]:
arr3 = np.concatenate((arr1,arr2))
arr3

array([   10,    20,    30,    40,    50,    60,    70,    80,    90,
         100,  1000,  2000,  3000,  4000,  5000,  6000,  7000,  8000,
        9000, 10000])

## Q6) Use array created in Q5 and replace the values lesser than 50 as <50.

In [158]:
np.where(arr3<50,'<50',arr3)

array(['<50', '<50', '<50', '<50', '50', '60', '70', '80', '90', '100',
       '1000', '2000', '3000', '4000', '5000', '6000', '7000', '8000',
       '9000', '10000'], dtype='<U11')

## Q7) Write a Code to create 1D, 2D and 3D arrays randomly.

In [130]:
np.random.rand(5) ## 1D

array([0.60054892, 0.9501295 , 0.23030288, 0.54848992, 0.90912837])

In [131]:
np.random.rand(5,4) ## 2D

array([[0.13316945, 0.52341258, 0.75040986, 0.66901324],
       [0.46775286, 0.20484909, 0.49076589, 0.37238469],
       [0.47740115, 0.36589039, 0.83791799, 0.76864751],
       [0.31399468, 0.57262533, 0.27604905, 0.45284293],
       [0.35297837, 0.65739946, 0.37035108, 0.45909298]])

In [150]:
np.random.rand(3,2,2) ## 3D

array([[[0.09714648, 0.10284749],
        [0.7015073 , 0.89047987]],

       [[0.1595603 , 0.27557254],
        [0.67249153, 0.16430312]],

       [[0.70137114, 0.48763522],
        [0.68067777, 0.52154819]]])

## Q8) Write a Code to create an array with 10 random integers within 100.

In [138]:
np.random.randint(0,100,10)

array([47, 16, 12, 93,  9, 76, 67, 14, 20, 53])

## Q9) Create a array with an sample data with shape of (5,4) with standard normal distribution with random seed of 201.

In [151]:
np.random.seed(201)
arr = np.random.normal(loc=0,scale=1,size=(5,4))

In [152]:
arr

array([[ 0.31678346, -0.4934171 ,  1.21794581,  0.43677076],
       [-0.72769267,  0.43030744, -0.68666532, -0.45301801],
       [-0.73285921, -0.75362743, -0.83251892,  0.44987187],
       [ 1.80752304, -0.48953136,  0.88158874, -0.69834087],
       [-0.0499747 ,  1.90752236, -0.32762267,  1.15003185]])

## Q10) Use the array from Q9 and create a dataframe with index=['L','M','N','O','P'], columns=['A','B','C','D']

In [153]:
pd.DataFrame(arr,index=['A','B','C','D','E'],columns=['W','X','Y','Z'])

W         X         Y         Z
A  0.316783 -0.493417  1.217946  0.436771
B -0.727693  0.430307 -0.686665 -0.453018
C -0.732859 -0.753627 -0.832519  0.449872
D  1.807523 -0.489531  0.881589 -0.698341
E -0.049975  1.907522 -0.327623  1.150032

## Q11) Find the most frequent value in an numpy 1D array. Use array x = np.array([10,20,66,20,10,30,50,60,70,60,66,12,13,66])

In [163]:
x = np.array([10,20,66,20,10,30,50,60,70,60,66,12,13,66])
print(x)
  
print("Most frequent value in the array is:",np.bincount(x).argmax())

[10 20 66 20 10 30 50 60 70 60 66 12 13 66]
Most frequent value in the array is: 66


## Q12) Write a Code to Flatten a 2d array into 1d array. Use array: np.array([[1,2,3],[8,9,10],[11,22,33]])

In [168]:
a1 = np.array([[1,2,3],[8,9,10],[11,22,33]])
a1

array([[ 1,  2,  3],
       [ 8,  9, 10],
       [11, 22, 33]])

In [169]:
a1.flatten()

array([ 1,  2,  3,  8,  9, 10, 11, 22, 33])

<a id="pd"> </a>
# *Part II: PANDAS*

## About Dataset Provided.

- **YouTube Dislikes Dataset:**
This dataset contains information about trending YouTube videos from August 2020 to December 2021 for the USA, Canada, and Great Britain.
- **Context and Content:**
Youtube announced the decision to hide the number of dislikes from users around November 2021. However, the official YouTube Data API allowed you to get information about dislikes until December 13, 2021.
- This dataset contains the latest possible information about dislikes,likes,views and more which was collected just before December 13. The information was collected by videos that had been trending in the USA, Canada, and Great Britain for a year prior.


## Q1) Import required libraries and read the provided dataset (youtube_dislike_dataset.csv) and retrieve top 5 and bottom 5 records.

In [16]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [18]:
df = pd.read_csv('youtube_dislike_dataset.csv')
df.head(5)

video_id                                       title  \
0  --0bCF-iK2E        Jadon Sancho  Magical Skills & Goals   
1  --14w5SOEUs          Migos - Avalanche (Official Video)   
2  --40TEbZ9Is  Supporting Actress in a Comedy: 73rd Emmys   
3  --4tfbSyYDE     JO1'YOUNG (JO1 ver.)' PERFORMANCE VIDEO   
4  --DKkzWVh-E                Why Retaining Walls Collapse   

                 channel_id          channel_title         published_at  \
0  UC6UL29enLNe4mqwTfAyeNuw             Bundesliga  2021-07-01 10:00:00   
1  UCGIelM2Dj3zza3xyV3pL3WQ              MigosVEVO  2021-06-10 16:00:00   
2  UClBKH8yZRcM4AsRjDVEdjMg     Television Academy  2021-09-20 01:03:32   
3  UCsmXiDP8S40uBeJYxvyulmA                    JO1  2021-03-03 10:00:17   
4  UCMOqf8ab-42UUQIdVoKwjlQ  Practical Engineering  2021-12-07 13:00:00   

   view_count   likes  dislikes  comment_count  \
0     1048888   19515       226           1319   
1    15352638  359277      7479          18729   
2      925281   11212       401            831   
3     2641597   39131       441           3745   
4      715724   32887       367           1067   

                                                tags  \
0  football soccer ftbol alemn Bundesliga season ...   
1  Migos Avalanche Quality Control Music/Motown R...   
2                                                      
3  PRODUCE101JAPAN              JO1   TheSTAR STA...   
4  retaining wall New Jersey highway Direct Conne...   

                                         description  \
0  Enjoy the best skills and goals from Jadon San...   
1  Watch the the official video for Migos - "Aval...   
2  Hannah Waddingham wins the Emmy for Supporting...   
3  JO1'YOUNG (JO1 ver.)' PERFORMANCE VIDEO\n\n---...   
4  One of the most important (and innocuous) part...   

                                            comments  
0  Respect to Dortmund fans,must be sad losing hi...  
1  Migos just makes me want to live my live to th...  
2  Hannah's energy bursts through any screen. Wel...  
3  youngVer><  REN is really PERFECT. It's not ju...  
4   Keep up with all my projects here: https://pr...

In [19]:
df.tail(5)

video_id                                              title  \
37417  zzd4ydafGR0  Lil Tjay - Calling My Phone (feat. 6LACK) [Off...   
37418  zziBybeSAtw  PELICANS at LAKERS | FULL GAME HIGHLIGHTS | Ja...   
37419  zzk09ESX7e0                 [MV]  (MAMAMOO) - Where Are We Now   
37420  zzmQEb0Em5I              FELLIPE ESCUDERO- Master Podcast  #12   
37421  zzxPZwaA-8w  Gareth Bale brace secures dramatic comeback on...   

                     channel_id      channel_title         published_at  \
37417  UCEB4a5o_6KfjxHwNMnmj54Q           Lil Tjay  2021-02-12 05:03:49   
37418  UCWJ2lWNubArHWmf3FIHbfcQ                NBA  2021-01-16 05:39:05   
37419  UCuhAUMLzJxlP1W7mEk0_6lA            MAMAMOO  2021-06-02 09:00:10   
37420  UC8NjnNWMsRqq11NYvHAQb1g     Master Podcast  2020-10-20 20:59:30   
37421  UCEg25rdRZXg32iwai6N6l0w  Tottenham Hotspur  2021-05-23 21:00:31   

       view_count    likes  dislikes  comment_count  \
37417   120408275  2180780     35871          81360   
37418     2841917    20759      1049           2624   
37419    13346678   720854      4426          90616   
37420      252057    19198      1234           1471   
37421     2252090    34063       868           2004   

                                                    tags  \
37417  Lil Tjay Steady Calling My Phone Calling My Ph...   
37418  NBA G League Basketball game-0022000187 Lakers...   
37419  MAMAMOO  WAW  WAW MAMAMOO WAW Where Are We Now...   
37420  master masterpodcast lord lord vinheteiro z z ...   
37421  Spurs Tottenham Hotspur   Tottenham Leicester ...   

                                             description  \
37417  Official video for "Calling My Phone" by Lil T...   
37418  PELICANS at LAKERS | FULL GAME HIGHLIGHTS | Ja...   
37419  [MV]  (MAMAMOO) - Where Are We Now\n\nInstagra...   
37420  DOCTOR HAIR\nhttps://www.thedoctorhair.com/?fb...   
37421  Two minute highlights from Tottenham Hotspur's...   

                                                comments  
37417  'DESTINED 2 WIN' OUT NOW !! https://liltjay.ln...  
37418  Montrezl Harrell is going crazy with the rebou...  
37419  I honestly do not know why this song hit so ha...  
37420  Foi um prazer passar esta tarde com vocs debat...  
37421  Thank you Kane for everything you have given t...

## Q2) Check the info of the dataframe and write your inferences on datatypes and shape of the dataset.

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37422 entries, 0 to 37421
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   video_id       37422 non-null  object
 1   title          37422 non-null  object
 2   channel_id     37422 non-null  object
 3   channel_title  37422 non-null  object
 4   published_at   37422 non-null  object
 5   view_count     37422 non-null  int64 
 6   likes          37422 non-null  int64 
 7   dislikes       37422 non-null  int64 
 8   comment_count  37422 non-null  int64 
 9   tags           37422 non-null  object
 10  description    37422 non-null  object
 11  comments       37264 non-null  object
dtypes: int64(4), object(8)
memory usage: 3.4+ MB


In [21]:
df.shape

(37422, 12)

## Inference

- There are *12* columns and *37422* rows.
- There are *8* columns datatypes as object and rest all are integer datatype.
- We can consider column video_id as unique indentifer for all the rows.
- We can convert column published_at to pandas date_time format.
- We can observe some null values in the column comments.

## Q3) Check for the Percentage of the Null values and drop it.

In [31]:
round(df.isnull().sum()/len(df)*100,2)

video_id         0.00
title            0.00
channel_id       0.00
channel_title    0.00
published_at     0.00
view_count       0.00
likes            0.00
dislikes         0.00
comment_count    0.00
tags             0.00
description      0.00
comments         0.42
dtype: float64

* We can see there is *0.42%* of records in column records is null.

In [32]:
df = df.dropna()

In [33]:
df.shape

(37264, 12)

## Q4) Check the statistical summary of the dataset.

In [35]:
df.describe()

view_count         likes      dislikes  comment_count
count  3.726400e+04  3.726400e+04  3.726400e+04   3.726400e+04
mean   5.710821e+06  1.672867e+05  4.996434e+03   9.966953e+03
std    2.431304e+07  5.384931e+05  3.075194e+04   1.173465e+05
min    2.036800e+04  2.200000e+01  3.000000e+00   1.000000e+00
25%    5.141305e+05  1.331725e+04  2.827500e+02   9.140000e+02
50%    1.321860e+06  4.255600e+04  7.980000e+02   2.347000e+03
75%    3.674527e+06  1.309665e+05  2.466250e+03   6.212000e+03
max    1.322797e+09  3.183768e+07  2.397733e+06   1.607103e+07

## Q5) Convert datatype of column published_at from object to pandas datetime.

In [42]:
df['published_at'] = pd.to_datetime(df['published_at'])

In [49]:
df[['published_at']].info()  ## We can see that we sucessfully changed the dataype from object to datetime.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37264 entries, 0 to 37421
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   published_at  37264 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 582.2 KB


## Q6) Create a column as 'published_month' using the column published_at.

In [52]:
df['published_month'] = df['published_at'].dt.month

In [55]:
df.head(2)  ## We can see that we successfully created another column published_month.

video_id                                 title  \
0  --0bCF-iK2E  Jadon Sancho  Magical Skills & Goals   
1  --14w5SOEUs    Migos - Avalanche (Official Video)   

                 channel_id channel_title        published_at  view_count  \
0  UC6UL29enLNe4mqwTfAyeNuw    Bundesliga 2021-07-01 10:00:00     1048888   
1  UCGIelM2Dj3zza3xyV3pL3WQ     MigosVEVO 2021-06-10 16:00:00    15352638   

    likes  dislikes  comment_count  \
0   19515       226           1319   
1  359277      7479          18729   

                                                tags  \
0  football soccer ftbol alemn Bundesliga season ...   
1  Migos Avalanche Quality Control Music/Motown R...   

                                         description  \
0  Enjoy the best skills and goals from Jadon San...   
1  Watch the the official video for Migos - "Aval...   

                                            comments  published_month  
0  Respect to Dortmund fans,must be sad losing hi...                7  
1  Migos just makes me want to live my live to th...                6

## Q7) Replace the numbers in column published_month as names of the months i,e., 1 as 'Jan', 2 as 'Feb' and so on.....

In [57]:
df['published_month'] = df['published_month'].replace({1:'Jan',2:'Feb',3:'March',4:'April',5:'May',6:'June',
                                                      7:'July',8:'Aug',9:'Sep',10:'Oct',11:'Nov',12:'Dec'})

In [64]:
df['published_month'].head(4)

0     July
1     June
2      Sep
3    March
Name: published_month, dtype: object

## Q8) Find the minimum and maximum value of the view_count, likes, dislikes and comment_count and write your inferences.

In [76]:
df[['view_count','likes','dislikes','comment_count']].min()

view_count       20368
likes               22
dislikes             3
comment_count        1
dtype: int64

In [77]:
df[['view_count','likes','dislikes','comment_count']].max()

view_count       1322796924
likes              31837675
dislikes            2397733
comment_count      16071029
dtype: int64

## Inference.

- We can see that the mimimum view_count is *20368* and max is *1322796924.*
- The mimimum likes is *22* and max is *31837675.*
- The mimimum dislikes is *3* and max is *2397733.*
- The mimimum comment_count is *1* and max is *16071029.*

## Q9) Find the count of unique video_id, channel_id and channel_title.

In [80]:
df['video_id'].nunique()

37264

In [82]:
df['channel_id'].nunique()

10891

In [84]:
df['channel_title'].nunique()

10813

## Q10) Find the top10 channel names having the highest number of videos in the dataset and the bottom10 having lowest number of videos.

In [89]:
df.groupby('channel_title')['video_id'].count().sort_values(ascending=False).head(10)

channel_title
Sky Sports Football    533
The United Stand       301
BT Sport               246
NBA                    209
NFL                    162
WWE                    122
SSSniperWolf            99
SSundee                 98
FORMULA 1               87
NHL                     86
Name: video_id, dtype: int64

In [90]:
df.groupby('channel_title')['video_id'].count().sort_values(ascending=True).head(10)

channel_title
   SilverName                   1
Miniatur Wunderland             1
MiniBloxia                      1
Mini Muka                       1
Mini Ladd                       1
MindYourLogic                   1
Mind Body Tonic With Dr Sita    1
Mimi Ar                         1
Millyz                          1
Milkair                         1
Name: video_id, dtype: int64

## Q11) Find the title of the video which has the maximum number of likes and the title of the video having minimum likes and write your inferences.

In [101]:
pd.pivot_table(df,index=['title'],values=['likes'],aggfunc=np.max).sort_values(by='likes',ascending=False).head(1)

likes
title                                  
BTS () 'Dynamite' Official MV  31837675

##### We can see that 'BTS Dynamite Offical Music Video' is having the highest number of likes.

In [102]:
pd.pivot_table(df,index=['title'],values=['likes'],aggfunc=np.max).sort_values(by='likes',ascending=True).head(1)

likes
title                               
Leading the Charge | Circle K     22

##### 'Leading the charge | Circle K' is having lowest number of likes.

## Q12) Find the title of the video which has the maximum number of dislikes and the title of the video having minimum dislikes and write your inferences.

In [107]:
pd.pivot_table(df,index=['title'],values=['dislikes'],aggfunc=np.max).sort_values(by='dislikes',ascending=False).head(1)

dislikes
title                                        
Cuties | Official Trailer | Netflix   2397733

##### 'Cuties | Offical Trailer | Netflix' has the highest number of dislikes.

In [108]:
pd.pivot_table(df,index=['title'],values=['dislikes'],aggfunc=np.max).sort_values(by='dislikes',ascending=True).head(1)

dislikes
title                                           
Tims For Good: A Taste Of The Familiar         3

##### 'Tims For Good: A Taste Of The Familiar' has the lowest number of dislikes.